README.md
# TenLog
## A lightweight logger

[![CocoaPods](https://img.shields.io/cocoapods/v/TenLog.svg)]()
[![license](https://img.shields.io/github/license/mitch10e/TenLog.svg)]()

[![GitHub stars](https://img.shields.io/github/stars/mitch10e/TenLog.svg?style=social&label=Star)]()
[![Twitter Follow](https://img.shields.io/twitter/follow/mitchell10e.svg?style=social&label=Follow)]()

TenLog is an alternative to using the default ```print()``` statement in your Swift projects, 
Created and maintained by [Mitchell Tenney](https://gitlab.com/u/mitch10e)

## Installation

### CocoaPods

[CocoaPods](http://cocoapods.org) is a dependency manager for Cocoa projects. You can install it with the following command:

To integrate TenLog into your Xcode project using CocoaPods, specify it in your `Podfile`:

```ruby
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '10.0'
use_frameworks!

target '<Your Target Name>' do
    pod 'TenLog', '~> 1.0.0'
end
```

Then, run the following command:

```bash
$ pod install
```

## Usage

To use TenLogs, simply use the ```debug()``` statement as if it were ```print()```

You can change the settings of the output via the ```TenLogSettings``` class.

```ruby
debug("Hello, World!")
debug("How are you?\nI am fine.")
debug("Do you like to have organized logs in your project?\nYes! We hate the timestamp stuff!\nWe usually don't even read the timestamps!\nAnd what is it with those random AppName[Numbers:Numbers] stuff?\nAverage developers don't care about this stuff!\nWe just want something easy to scan through!")

debug("http://www.example.com", tag: "get")
debug("http://www.example.com", tag: "200")

debug("http://www.failurl.com", tag: "get")
debug("http://www.failurl.com", tag: "404")

TenLogSettings.disable(tag: "get")
TenLogSettings.disable(tag: "200")
TenLogSettings.disable(tag: "404")
debug("http://www.example.com", tag: "get")
debug("http://www.example.com", tag: "200")

debug("http://www.failurl.com", tag: "get")
debug("http://www.failurl.com", tag: "404")


let tableFlip = "(╯°□°）╯︵ ┻━┻"
debug(tableFlip, tag: "rage")

debug("Just use debug() like a normal print()!", tag: "Info")
debug("Check out the TenLogSettings to modify how the logs work!", tag: "Info")

debug("Item 1", "Item 2", "Item 3", tag: "Test")

debug("Don't forget to ⭐️ the repo!\n\tThank You!", tag: "Vital")
```

will output the following:

```ruby
DEBUG  | Hello, World!
DEBUG  | How are you?
       | I am fine.
       ˇ
DEBUG  | Do you like to have organized logs in your project?
       | Yes! We hate the timestamp stuff!
       | We usually don't even read the timestamps!
       | And what is it with those random AppName[Numbers:Numbers] stuff?
       | Average developers don't care about this stuff!
       | We just want something easy to scan through!
       ˇ
GET    | http://www.example.com
200    | http://www.example.com
GET    | http://www.failurl.com
404    | http://www.failurl.com
RAGE   | (╯°□°）╯︵ ┻━┻
INFO   | Just use debug() like a normal print()!
INFO   | Check out the TenLogSettings to modify how the logs work!
TEST   | Item 1
TEST   | Item 2
TEST   | Item 3
VITAL  | Don't forget to ⭐️ the repo!
       | 	Thank You!
       ˇ
```



## License
- This library uses the MIT License
  - https://opensource.org/licenses/mit-license.html

## Changelog

[v1.0.0](https://github.com/mitch10e/TenLog/blob/develop/CHANGELOG.md#100)

[v0.2.2](https://github.com/mitch10e/TenLog/blob/develop/CHANGELOG.md#022)

[v0.1.0](https://github.com/mitch10e/TenLog/blob/develop/CHANGELOG.md#010)

## Versioning

TenLog uses the Semantic Versioning guidelines as much as possible. Releases will be numbered with the following format:

`<major>.<minor>.<patch>`

For more information on Semantic Versioning, please visit http://semver.org.

## Author
- Email: mitch10e@gmail.com
- Twitter: http://twitter.com/mitchell10e
- GitHub: https://github.com/mitch10e
- GitLab: https://gitlab.com/u/mitch10e
- LinkedIn: https://www.linkedin.com/in/mitch10e



CHANGELOG.md
# TenLog CHANGELOG

## 1.0.0

###### Enhancements

- Moved to GitHub. [TenLog Repository](https://github.com/mitch10e/TenLog)
- Added Shields to README. [Sheilds.io](https://shields.io/)
- Ready to launch 1.0.0!

## 0.2.2

###### Enhancements

- Added functionality to disable debug statements by tag

## 0.1.0

- Created initial library.


Source/Classes/TenLog.swift
//
//  TenLog.swift
//  TenLog
//
//  Created by Mitchell Tenney on 10/7/16.
//  Copyright © 2016 Mitchell Tenney. All rights reserved.
//

import Foundation

/**
 The core of the TenLog library. debug() can be used exactly as the print() function. Acts as a pretty print function.
 - Parameter items: What to print out
 - Parameter tag: Override the default tag
 */

public func debug(_ items: Any..., tag: String = "\(TenLogSettings.shared.defaultTag)") {
    if shouldPrint(tag) {
        for item in items {
            print("\(normalize(tag))\(TenLogSettings.shared.separator)\(normalize(item))")
        }
    }
}

fileprivate func shouldPrint(_ tag: String) -> Bool {
    var shouldPrint = true
    if !TenLogSettings.shared.enabled {
        shouldPrint = false
    }
    if TenLogSettings.shared.disabledTags.contains(tag.uppercased()) {
        shouldPrint = false
    }    
    return shouldPrint
}

fileprivate func normalize(_ item: Any) -> String {
    var value = "\(item)"
    if value.contains("\n") {
        let spaces = String(repeating: " ", count: TenLogSettings.shared.tagSize)
        value = value.replacingOccurrences(of: "\n", with: "\n\(spaces)\(TenLogSettings.shared.separator)")
        value += "\n\(getSpaces(TenLogSettings.shared.tagSize))\(TenLogSettings.shared.endMultiLineCharacter)"
    }
    return value
}

fileprivate func normalize(_ tag: String) -> String {
    var normalizedTag = (String(tag.characters.prefix(TenLogSettings.shared.tagSize)))
    let spaceLeft = TenLogSettings.shared.tagSize - normalizedTag.characters.count
    let spaces = getSpaces(spaceLeft)
    
    if TenLogSettings.shared.alignment == .left {
        normalizedTag = normalizedTag + spaces
    } else {
        normalizedTag = spaces + normalizedTag
    }
    if TenLogSettings.shared.shouldUppercase {
        normalizedTag = normalizedTag.uppercased()
    }
    return normalizedTag
}

fileprivate func getSpaces(_ count: Int) -> String {
    return String(repeating: " ", count: count)
}


Source/Classes/TenLogSettings.swift
//
//  TenLogSettings.swift
//  TenLog
//
//  Created by Mitchell Tenney on 10/7/16.
//  Copyright © 2016 Mitchell Tenney. All rights reserved.
//

import Foundation

public enum TenLogAlignment {
    case left
    case right
}


public class TenLogSettings {
    
    // Private init - Singleton
    private init() {}
    static let shared: TenLogSettings = TenLogSettings()
    
    
    // MARK: Internal values for settings
    
    var tagSize: Int = 7 {
        didSet {
            if tagSize < defaultTag.characters.count {
                tagSize = defaultTag.characters.count
                debug("tagSize smaller than default tag. tagSize overriding to \(tagSize)")
            }
        }
    }
    
    var defaultTag: String = "DEBUG" {
        didSet {
            if defaultTag.characters.count > TenLogSettings.shared.tagSize {
                TenLogSettings.setTagSize(defaultTag.characters.count)
            }
        }
    }
    
    var enabled: Bool = true
    var alignment: TenLogAlignment = .left
    var endMultiLineCharacter: Character = "ˇ"
    var separator: String = "| "
    var shouldUppercase: Bool = true
    var disabledTags: Set<String> = []
    
    
    
    // MARK: Public API
    
    /**
     Enable debug() print statements
     */
    
    public static func enable() {
        TenLogSettings.shared.enabled = true
    }
    
    /**
     Enables a tag that has been previously disabled. Will uppercase tag to remove automagically.
     - Parameter tag: The tag to enable
     */
    
    public static func enable(tag: String) {
        TenLogSettings.shared.disabledTags.remove(tag.uppercased())
    }
    
    /**
     Disable debug() print statements
     */
    
    public static func disable() {
        TenLogSettings.shared.enabled = false
    }
    
    /**
     Disables a tag from being printed via debug(). Stored as uppercased().
     - Parameter tag: The tag to disable
     */
    
    public static func disable(tag: String) {
        TenLogSettings.shared.disabledTags.insert(tag.uppercased())
    }
    
    /**
     Get a list of all tags currently disabled
     */
    
    public static func getDisabledTags() -> Set<String> {
        return TenLogSettings.shared.disabledTags
    }
    
    /**
     Pass in a set of tags to disable
     - Parameter tags: List of tags to add to disabled tag set
     */
    
    public static func disable(tags: [String])  {
        for tag in tags {
            TenLogSettings.shared.disabledTags.insert(tag)
        }
    }
    
    /**
     Enables all tags
     */
    
    public static func enableAllTags() {
        TenLogSettings.shared.disabledTags.removeAll()
    }
    
    /**
     Sets the number of characters that should come before the separator
     - Parameter size: The size of the tag. Defaults to 7
     */
    
    public static func setTagSize(_ size: Int) {
        TenLogSettings.shared.tagSize = size
    }
    
    /**
     Sets the default string to use when no tag is specified in debug(). Will automatically update the tag size to the length of the default tag if tagSize is smaller than the tag passed in.
     - Parameter tag: The default string to use as a tag. Defaults to DEBUG
     */
    
    public static func setDefaultTag(_ tag: String) {
        TenLogSettings.shared.defaultTag = tag
    }
    
    /**
     Sets the tag to be left or right aligned. Use TenLogAlignment{ .left, .right }
     - Parameter alignment: Text alignment to use for tags in debug(). Defaults to .left
     */
    
    public static func setAlignment(_ alignment: TenLogAlignment) {
        TenLogSettings.shared.alignment = alignment
    }
    
    /**
     When multiline values with \\n are passed into debug() an extra line is created with this character to help define end of statement.
     - Parameter char: The character to use as end of line. Defaults to ˇ
     */
    
    public static func setEndMultiLineCharacter(_ char: Character) {
        TenLogSettings.shared.endMultiLineCharacter = char
    }
    
    /**
     A separator is used to help divide tags from items being printed. Use this to change to something else. Common options to change to are empty string or a character that lines up well.
     - Parameter separator: The string to use to separate the tag from the items. Defaults to "| "
     */
    
    public static func setSeparator(_ separator: String) {
        TenLogSettings.shared.separator = separator
    }
    
    /**
     For the more lazy out there, this will tell the tags to always uppercased() the tag, even if "debug" is set as the tag.
     - Parameter uppercase: Defaults to true
     */
    
    public static func setShouldUppercase(_ uppercase: Bool) {
        TenLogSettings.shared.shouldUppercase = uppercase
    }
}


Example/TenLog/TenLog/AppDelegate.swift
//
//  AppDelegate.swift
//  TenLog
//
//  Created by Mitchell Tenney on 10/7/16.
//  Copyright © 2016 Mitchell Tenney. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



Example/TenLog/TenLogTests/TenLogTests.swift
//
//  TenLogTests.swift
//  TenLogTests
//
//  Created by Mitchell Tenney on 10/7/16.
//  Copyright © 2016 Mitchell Tenney. All rights reserved.
//

import XCTest
@testable import TenLog

class TenLogTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }
    
}


Example/TenLog/TenLog/ViewController.swift
//
//  ViewController.swift
//  TenLog
//
//  Created by Mitchell Tenney on 10/7/16.
//  Copyright © 2016 Mitchell Tenney. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        debug("Hello, World!")
        debug("How are you?\nI am fine.")
        debug("Do you like to have organized logs in your project?\nYes! We hate the timestamp stuff!\nWe usually don't even read the timestamps!\nAnd what is it with those random AppName[Numbers:Numbers] stuff?\nAverage developers don't care about this stuff!\nWe just want something easy to scan through!")

        debug("http://www.example.com", tag: "get")
        debug("http://www.example.com", tag: "200")
        
        debug("http://www.failurl.com", tag: "get")
        debug("http://www.failurl.com", tag: "404")
        
        TenLogSettings.disable(tag: "get")
        TenLogSettings.disable(tag: "200")
        TenLogSettings.disable(tag: "404")
        debug("http://www.example.com", tag: "get")
        debug("http://www.example.com", tag: "200")
        
        debug("http://www.failurl.com", tag: "get")
        debug("http://www.failurl.com", tag: "404")
        
        
        let tableFlip = "(╯°□°）╯︵ ┻━┻"
        debug(tableFlip, tag: "rage")

        debug("Just use debug() like a normal print()!", tag: "Info")
        debug("Check out the TenLogSettings to modify how the logs work!", tag: "Info")
        
        debug("Item 1", "Item 2", "Item 3", tag: "Test")
        
        debug("Don't forget to ⭐️ the repo!\n\tThank You!", tag: "Vital")
    }
    
}

